Support strings in the Index checker

panacekcz opened the pull request:
This is a second part of the Bounded-size strings GSoC project. I submit this PR to start a review process. Could you please review the changes, @kelloggm?

Summary of changes:
- in the index checker, treat `string.length()` the same as `array.length`
- enable and change annotations on string indices in the index JDK
- enable annotations on `CharSequence` indices in the index JDK. Only preconditions are enabled, in order to avoid getting index qualifiers on mutable sequences such as `StringBuilder`.
- refer to "sequences" instead of "arrays" in the manual, comments, and variable and method names

Remaining issues:
- test `CheckAgainstNegativeOne` has been disabled, see panacekcz#4
- ~~this PR ignores the fact that `String` implements `CharSequence`. See panacekcz#3 for details and possible approaches. All comments and questions are welcome there. The current situation corresponds to solution 2.~~
- ~~`.travis.yml` is edited to build the jdk~~
- ~~should not be merged before #1352~~
- `build.xml` is edited to download the annotated jdk built from this code

panacekcz commented:
Junit tests fail on Travis, because one test requires the changes done in #1352, which are not included in this PR.
They pass in a [branch](https://github.com/panacekcz/checker-framework/tree/bounded-strings-index-dev) which combines this and #1352: https://travis-ci.org/panacekcz/checker-framework/builds/255544911

panacekcz commented:
I have tested the changes on on [plume-lib](https://github.com/mernst/plume-lib-check-index-new/tree/3f82f73faaf4c01bee92944a1ea16d8e854bf2fb). I created a version with suppressed warnings and links to issues: https://github.com/mernst/plume-lib-check-index-new/compare/master...panacekcz:master

The issues encountered are:
- #1411: Index checker crash (1 occurence).
- kelloggm#165: Upper bound not inferred for a string literal (1 occurence).
- kelloggm#56: Calling `substring` after `startsWith` or `endsWith` returns `true` (12 occurences).
- panacekcz#4: Calling `substring` with an index returned from `indexOf` plus the length of the string (6 occurences).
- panacekcz#5: Calling `substring` with end `offset+length` (1 occurence). Calling `write(String,int,int)` with offset `0` (1 occurence). 
- kelloggm#164: Warning on clone method, not related to strings (2 occurences).
- 1 [known bug](https://github.com/mernst/plume-lib-check-index-new/blob/3f82f73faaf4c01bee92944a1ea16d8e854bf2fb/java/src/plume/SimpleLog.java#L188) calling `substring`
- 1 [call to `substring`](https://github.com/mernst/plume-lib-check-index-new/blob/3f82f73faaf4c01bee92944a1ea16d8e854bf2fb/java/src/plume/ICalAvailable.java#L534) I don't know if it is correct or a bug

kelloggm commented on checker/src/org/checkerframework/checker/index/upperbound/UpperBoundAnnotatedTypeFactory.java:
```diff
@@ -76,7 +76,7 @@
     public final AnnotationMirror UNKNOWN, BOTTOM, POLY;
 
-    private final IndexMethodIdentifier imf;
+    protected final IndexMethodIdentifier imf;
```
Why is this `protected` instead of package-private (no modifier)? It's not used anywhere outside the package, and package-private prevents subclasses outside the package from seeing the field. I'm not sure why anyone would want to subclass `UpperBoundAnnotatedTypeFactory`, but we can expose this if there's a good reason.

panacekcz commented on checker/src/org/checkerframework/checker/index/lowerbound/LowerBoundAnnotatedTypeFactory.java:
```diff
@@ -309,21 +309,39 @@ public Void visitMethodInvocation(MethodInvocationTree tree, AnnotatedTypeMirror
+         * Looks up the minlen of a mehtod invocation tree. Returns null if the tree doesn't
+         * represent an string length method.
+         */
+        private Integer getMinLenFromMethodInvocationTree(MethodInvocationTree tree) {
```
This would mean the method would have 4 parameters `(MethodInvocationTree, ValueAnnotatedTypeFactory, IndexMethodIdentifier, ProcessingEnvironment)`. I think it would be less readable, given how short the method is. Is it worth it?

kelloggm replied:
That's a good point. I'll defer to your judgment.

panacekcz commented:
@kelloggm, I have addressed the review comments. Also, I enabled annotations for `CharSequence` to avoid unsoundness when accessing string through that interface.

kelloggm commented on checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java:
```diff
@@ -322,7 +322,7 @@ private void refineNeqSequenceLength(
             receiver = fa.getReceiver();
 
-        } else if (IndexUtil.isStringLengthInvocation(lengthAccess, atypeFactory.imf)) {
+        } else if (atypeFactory.imf.isStringLengthInvocation(lengthAccess)) {
```
I made a similar comment on panacekcz#6, but this looks like a code smell because you're reading through a field. You should add a getter to the type factory that just returns the Index Method Identifier and use it here.

panacekcz commented:
I have addressed the review comment. Test should pass after #1352 is merged.

kelloggm commented:
The tests appear to be failing due to timing out. I restarted the ones that timed out (sometimes they'll pass the second time), but if this continues we'll have to either up the timeout or look into whether there's a significant performance regression in this PR somewhere.

mernst commented:
Even if re-running these tests multiple times manages to get them to finish under 50 minutes, pushing the changes would mean that master would have these same recurring timeout issues.  Therefore, restarting the tests can help determine whether there is a functionality error, but does not enable the pull request to be merged.

The Index Checker already imposes a slowdown, and adding it to the Checker Framework is the reason that Checker Framework Travis jobs already sometimes time out.  Adding new functionality beyond that make take us over the 50-minute timeout.  Two possibilities are to profile performance (both in the Index Checker and is these proposed changes) or to split the Travis jobs into smaller pieces.

panacekcz commented:
The `.travis.yml` file in this PR is changed to build the jdk, because there are changes to the index jdk. So I think that may cause the timeouts. I wouldn't expect a massive slowdown due to this PR, but I don't know how to test it on travis in this situation.

mernst commented:
Rather than building the JDK, it's better to prebuild it, put it somewhere, and then download it in your fork.
You can do this, or @smillst can help with this.

panacekcz commented:
I published the jdk built from this branch and changed the url to point at it. The Travis build is now passing.
Changes to `build.xml` should be reverted before merging.

mernst commented:
Great, thanks.
Also, before merging the canonical prebuilt JDK should be updated to reflect your changes.

smillst commented:
FYI, I'm making the August release now, so I'll merge this later today or tomorrow.

smillst commented:
I'm rebuilding the JDK to upload to checker-framework.org/dev-jdk; once that's done I'll merge.
